src/components/utils.ts

```diff
@@ -0,0 +1,33 @@
+class ManagedProps<T> {
+  private usedProps: (keyof T)[] = [];
+  private rawProps: T;
+
+  constructor(props: T) {
+    this.rawProps = props;
+
+    for (const key in props) {
+      // Define getters and setters dynamically
+      Object.defineProperty(this, key, {
+        get: () => {
+          this.usedProps.push(key);
+          return this.rawProps[key];
+        },
+        enumerable: true,
+      });
+    }
+  }
+
+  get remainingProps(): T {
+    const mRemainingProps: Record<string | number | symbol, unknown> = {};
```

Comment:
In this case, it's not a problem, since `remainingProps` is read exactly once but ideally, this would be memoized.